### PR TITLE
AND/OR compound triggers support

### PR DESCRIPTION
**Pull request by abdmob**
## What changes were proposed in this pull request?

Support for compound triggers.
For example:

```
optimizer.setEndWhen(
   Trigger.and(Trigger.maxScore(0.99f),Trigger.maxEpoch(10))
)
```

or

```
optimizer.setEndWhen(
   Trigger.or(Trigger.maxScore(0.99f),Trigger.maxEpoch(10))
)
```

## How was this patch tested?

Unit tests and manual tests

### CODE REVIEW

**Comment by qiuxin2012** [comment]
Looks cool, you can add the Unit test here https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/test/scala/com/intel/analytics/bigdl/optim/OptimizerSpec.scala

**Comment by abdmob** [comment]
done

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/optim/OptimizerSpec.scala [hunk 1]
```diff
@@ -151,6 +151,51 @@ class OptimizerSpec extends FlatSpec with Matchers with BeforeAndAfter {
...
+    falseDummyOptimizer.setEndWhen(Trigger.and(trueDummyTrigger, falseDummyTrigger))
+    falseDummyOptimizer.optimize()
+
+    falseDummyOptimizer.setEndWhen(Trigger.and(falseDummyTrigger, trueDummyTrigger))
+    falseDummyOptimizer.optimize()
+
+    falseDummyOptimizer.setEndWhen(Trigger.and(falseDummyTrigger, falseDummyTrigger))
+    falseDummyOptimizer.optimize()
+
+    // OR
```
Comment (qiuxin2012) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/optim/OptimizerSpec.scala]: A true true is missing?

**Comment by qiuxin2012** [comment]
@abdmob Style check failed, see errors below:
```
[INFO] --- scalastyle-maven-plugin:0.8.0:check (default) @ bigdl ---
error file=/opt/work/jenkins/workspace/BigDL-PRVN-StyleCheck/spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Trigger.scala message=Use Javadoc style indentation for multiline comments line=127 column=0
error file=/opt/work/jenkins/workspace/BigDL-PRVN-StyleCheck/spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Trigger.scala message=Use Javadoc style indentation for multiline comments line=140 column=0
```

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Trigger.scala [hunk 2]
```diff
@@ -123,5 +123,33 @@ object Trigger {
       }
     }
   }
+
+  /**
+    * A trigger contains other triggers and triggers when all of them trigger (logical AND)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Trigger.scala]: indent here

**Comment by qiuxin2012** [comment]
jenkins passed

**Comment by qiuxin2012** [comment]
fix #2675

**Comment by qiuxin2012** [comment]
fix #2672

**Pull request merged**